File: core/src/main/java/com/mcreater/amclcore/resources/ResourceFetcher.java
package com.mcreater.amclcore.resources;

import lombok.Getter;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class ResourceFetcher {
    @Getter
    private static URLClassLoader CLASSLOADER = new URLClassLoader(new URL[0], ResourceFetcher.class.getClassLoader());

    public static void addRes(URL[] urls) throws IOException {
        URL[] u = CLASSLOADER.getURLs();
        CLASSLOADER.close();
        CLASSLOADER = null;
        CLASSLOADER = new URLClassLoader(Stream.concat(Arrays.stream(u), Arrays.stream(urls)).toArray(URL[]::new), ResourceFetcher.class.getClassLoader());
    }

    public static void addRes(URL url) throws IOException {
        addRes(new URL[]{url});
    }

    public static void addRes(List<URL> url) throws IOException {
        addRes(url.toArray(new URL[0]));
    }

    public static void removeRes(URL[] urls) throws IOException {
        URL[] u = CLASSLOADER.getURLs();
        List<URL> u2 = Arrays.stream(u).collect(Collectors.toList());
        u2.removeAll(Arrays.stream(urls).collect(Collectors.toList()));
        CLASSLOADER.close();
        CLASSLOADER = null;
        CLASSLOADER = new URLClassLoader(u2.toArray(new URL[0]), ResourceFetcher.class.getClassLoader());
    }

    public static void removeRes(URL url) throws IOException {
        removeRes(new URL[]{url});
    }

    public static void removeRes(List<URL> url) throws IOException {
        removeRes(url.toArray(new URL[0]));
    }

    public static InputStream get(String id, String type, String name) {
        return get(String.format("assets/%s/%s/%s", id, type, name));
    }

    public static InputStream get(String path) {
        return CLASSLOADER.getResourceAsStream(path);
    }

    public static List<URL> getFiles(String path) throws IOException {
        return Collections.list(CLASSLOADER.getResources(path));
    }
}

File: core/src/main/java/com/mcreater/amclcore/resources/ResourceFetcher.kt
package com.mcreater.amclcore.resources

import java.io.IOException
import java.io.InputStream
import java.net.URL
import java.net.URLClassLoader
import java.util.*
import java.util.stream.Collectors
import java.util.stream.Stream


class ResourceFetcher {
    companion object {
        @JvmStatic
        private var CLASSLOADER: URLClassLoader? =
            URLClassLoader(arrayOfNulls(0), ResourceFetcher::class.java.getClassLoader())

        @JvmStatic
        fun classloader(): URLClassLoader? {
            return CLASSLOADER
        }

        @JvmStatic
        @Throws(IOException::class)
        fun addRes(urls: Array<URL>?) {
            val u = CLASSLOADER!!.urLs
            CLASSLOADER!!.close()
            CLASSLOADER = null
            CLASSLOADER = URLClassLoader(
                Stream.concat(Arrays.stream(u), Arrays.stream(urls)).toArray { arrayOf() },
                ResourceFetcher::class.java.getClassLoader()
            )
        }

        @JvmStatic
        @Throws(IOException::class)
        fun addRes(url: URL) {
            addRes(arrayOf(url))
        }

        @JvmStatic
        @Throws(IOException::class)
        fun addRes(url: List<URL>) {
            addRes(url.toTypedArray<URL>())
        }

        @JvmStatic
        @Throws(IOException::class)
        fun removeRes(urls: Array<URL>?) {
            val u = CLASSLOADER!!.urLs
            val u2 = Arrays.stream(u).collect(Collectors.toList())
            u2.removeAll(Arrays.stream(urls).collect(Collectors.toList()))
            CLASSLOADER!!.close()
            CLASSLOADER = null
            CLASSLOADER = URLClassLoader(u2.toTypedArray<URL>(), ResourceFetcher::class.java.getClassLoader())
        }

        @JvmStatic
        @Throws(IOException::class)
        fun removeRes(url: URL) {
            removeRes(arrayOf(url))
        }

        @JvmStatic
        @Throws(IOException::class)
        fun removeRes(url: List<URL>) {
            removeRes(url.toTypedArray<URL>())
        }

        @JvmStatic
        operator fun get(id: String?, name: String?): InputStream? {
            return ResourceFetcher[String.format("assets/%s/%s", id, name)]
        }

        @JvmStatic
        operator fun get(path: String?): InputStream? {
            return CLASSLOADER!!.getResourceAsStream(path)
        }

        @JvmStatic
        @Throws(IOException::class)
        fun getFiles(path: String?): List<URL> {
            return Collections.list(CLASSLOADER!!.getResources(path))
        }
    }
}
